http/src/main/scala/org/http4s/blaze/http/http2/PingManager.scala

```diff
@@ -1,4 +1,58 @@
 package org.http4s.blaze.http.http2
 
-// Place holder
-trait PingManager
+import java.nio.ByteBuffer
+
+import org.log4s.getLogger
+
+import scala.concurrent.{Future, Promise}
+import scala.concurrent.duration.Duration
+
+private class PingManager(session: SessionCore) {
+
+  private[this] case class PingState(startedSystemTimeMs: Long, continuation: Promise[Duration])
+
+  private[this] val logger = getLogger
+  private[this] var currentPing: Option[PingState] = None
+
+  // Must be called from within the sessionExecutor
+  def ping(): Future[Duration] = currentPing match {
+    case Some(_) =>
+      val msg = "Ping already in progress"
+      val ex = new IllegalStateException(msg)
+      logger.warn(ex)(msg)
```

Comment:
I'm not generally a huge fan of logging and passing an exception.  Makes it tend to appear twice when it happens once.

Reply:
Thats reasonable. Do you no like logging at debug level as well? I think this is debug worthy at a minimum, and it's not possible to control whether the passed exception actually gets logged.

Reply:
I think that's a fair compromise if you're worried about the caller not logging it.  I just tend think if you pass it you pass off responsibility, and get annoyed when I see twice as many exceptions as I have problems.

Reply:
This is only used privately so the details are completely in our hands here, so I'm convinced to not log here.